### PR TITLE
#202 Add static var compensator management

### LINKED ISSUE #202: Manage SVARC components

### PR DESCRIPTION
**Pull request by lecourtoisflo**

### CODE REVIEW

File: sources/Inputs/include/Behaviours.h [hunk 1]
```diff
@@ -122,5 +122,52 @@ struct ConverterInterface {
...
+
+/// @brief Static var compensator (SVarC) behaviour
+struct StaticVarCompensator {
+  using SVarCid = std::string;  ///< alias for static var compensator id
+
+  /**
+   * @brief Constructor
+   *
+   * @param id the id of the SVarC
+   * @param bMin the minimum b of the SVarC
```
Comment (marcochiaramello) [inline sources/Inputs/include/Behaviours.h:134]: bMin the minimum susceptance value of the SVarC
same below

File: sources/Inputs/include/Behaviours.h [hunk 2]
```diff
@@ -122,5 +122,52 @@ struct ConverterInterface {
...
+   * @param bMax the maximum b of the SVarC
+   * @param voltageSetPoint the voltage set point of the SVarC
+   * @param VNom the nominal voltage of the bus of the SVarC
+   * @param UMinActivation the low voltage threshold of the SVarC
```
Comment (marcochiaramello) [inline sources/Inputs/include/Behaviours.h:138]: the low voltage activation threshold of SVarC
similar below

File: sources/Inputs/include/Behaviours.h [hunk 3]
```diff
@@ -122,5 +122,52 @@ struct ConverterInterface {
...
+   * @param UMaxActivation the high voltage threshold of the SVarC
+   * @param USetPointMin the low voltage set point of the SVarC
+   * @param USetPointMax the high voltage set point of the SVarC
+   * @param b0 the initial b of the SVarC
+   * @param slope the slope of the SVarC
```
Comment (marcochiaramello) [inline sources/Inputs/include/Behaviours.h:143]: the slope (kV/MVar) of the SVarC

File: sources/Outputs/src/Par.cpp [hunk 4]
```diff
@@ -475,5 +486,51 @@ Par::updatePropParameters(boost::shared_ptr<parameters::ParametersSet> set) {
...
+  double value = computeBPU(svarc.bMax, svarc.VNom);
+  set->addParameter(helper::buildParameter("SVarC_BMaxPu", value));
+
+  value = computeBPU(svarc.bMin, svarc.VNom);
+  set->addParameter(helper::buildParameter("SVarC_BMinPu", value));
+
+  value = computeBPU(svarc.b0, svarc.VNom);
+  set->addParameter(helper::buildParameter("SVarC_BShuntPu", value));
+
+  value = svarc.slope * Sb_ / svarc.VNom;
```
Comment (marcochiaramello) [inline sources/Outputs/src/Par.cpp:505]: maybe you can create a method to evaluate slopePu, just to be similat to computeBPu

File: sources/Outputs/src/Par.cpp [hunk 5]
```diff
@@ -475,5 +486,51 @@ Par::updatePropParameters(boost::shared_ptr<parameters::ParametersSet> set) {
...
+
+boost::shared_ptr<parameters::MacroParameterSet>
+Par::writeMacroParameterSetStaticVarCompensators() {
+  auto macro = boost::make_shared<parameters::MacroParameterSet>(macroParameterSetStaticCompensator_);
+
+  macro->addReference(helper::buildReference("SVarC_Mode0", "regulatingMode", "INT"));
+  macro->addReference(helper::buildReference("SVarC_P0Pu", "p_pu", "DOUBLE"));
+  macro->addReference(helper::buildReference("SVarC_Q0Pu", "q_pu", "DOUBLE"));
+  macro->addReference(helper::buildReference("SVarC_U0Pu", "v_pu", "DOUBLE"));
+  macro->addReference(helper::buildReference("SVarC_UPhase0", "v_pu", "DOUBLE"));
```
Comment (marcochiaramello) [inline sources/Outputs/src/Par.cpp:526]: here it is a phase, not a voltage (angle_pu)

**Review by marcochiaramello — CHANGES_REQUESTED** [review]

File: sources/Inputs/include/Behaviours.h (hunk 1 above)
Comment (lecourtoisflo) [reply sources/Inputs/include/Behaviours.h:134]: OK

File: sources/Inputs/include/Behaviours.h (hunk 2 above)
Comment (lecourtoisflo) [reply sources/Inputs/include/Behaviours.h:138]: OK

File: sources/Inputs/include/Behaviours.h (hunk 3 above)
Comment (lecourtoisflo) [reply sources/Inputs/include/Behaviours.h:143]: OK

File: sources/Outputs/src/Par.cpp (hunk 4 above)
Comment (lecourtoisflo) [reply sources/Outputs/src/Par.cpp:505]: It is only needed here so I'm not sure it is usefull

File: sources/Outputs/src/Par.cpp (hunk 5 above)
Comment (lecourtoisflo) [reply sources/Outputs/src/Par.cpp:526]: OK

File: sources/Inputs/include/Behaviours.h [hunk 6]
```diff
@@ -122,5 +122,52 @@ struct ConverterInterface {
...
+   * @param id the id of the SVarC
+   * @param bMin the minimum susceptance value of the SVarC
+   * @param bMax the maximum susceptance value of the SVarC
+   * @param voltageSetPoint the voltage set point of the SVarC
+   * @param VNom the nominal voltage of the bus of the SVarC
+   * @param UMinActivation the low voltage activation threshold of the SVarC
+   * @param UMaxActivation the high voltage activation threshold of the SVarC
+   * @param USetPointMin the low voltage set point of the SVarC
+   * @param USetPointMax the high voltage set point of the SVarC
+   * @param b0 the initial b of the SVarC
```
Comment (marcochiaramello) [inline sources/Inputs/include/Behaviours.h:142]: the initial susceptance value of the SVarC

File: sources/Outputs/src/Par.cpp (hunk 4 above)
Comment (marcochiaramello) [reply sources/Outputs/src/Par.cpp:505]: ok

**Review by marcochiaramello — CHANGES_REQUESTED** [review]

File: sources/Inputs/include/Behaviours.h (hunk 6 above)
Comment (lecourtoisflo) [reply sources/Inputs/include/Behaviours.h]: OK

File: sources/Outputs/src/Dyd.cpp [hunk 7]
```diff
@@ -341,6 +365,13 @@ Dyd::writeMacroStaticRef() {
   ref->addStaticRef("state_value", "state");
   ret.push_back(ref);
 
+  ref = dynamicdata::MacroStaticReferenceFactory::newMacroStaticReference(macroStaticRefSVarCName_);
+  ref->addStaticRef("SVarC_PInjPu", "p");
+  ref->addStaticRef("SVarC_PInjPu", "q");
```
Comment (cossartque) [inline sources/Outputs/src/Dyd.cpp:370]: QInjPu

**Review by cossartque — CHANGES_REQUESTED** [review]

File: sources/Outputs/src/Dyd.cpp (hunk 7 above)
Comment (lecourtoisflo) [reply sources/Outputs/src/Dyd.cpp:370]: OK

File: tests/outputs/reference/TestDyd/TestDyd.dyd [hunk 8]
```diff
@@ -24,6 +27,12 @@
     <dyn:staticRef var="QPu_value" staticVar="q"/>
     <dyn:staticRef var="state_value" staticVar="state"/>
   </dyn:macroStaticReference>
+  <dyn:macroStaticReference id="StaticVarCompensatorStaticRef">
+    <dyn:staticRef var="SVarC_PInjPu" staticVar="p"/>
+    <dyn:staticRef var="SVarC_PInjPu" staticVar="q"/>
```
Comment (cossartque) [inline tests/outputs/reference/TestDyd/TestDyd.dyd:32]: QInjPu here too

File: tests/outputs/reference/TestDydDynModel/TestDydDynModel.dyd [hunk 9]
```diff
@@ -32,6 +35,12 @@
     <dyn:staticRef var="QPu_value" staticVar="q"/>
     <dyn:staticRef var="state_value" staticVar="state"/>
   </dyn:macroStaticReference>
+  <dyn:macroStaticReference id="StaticVarCompensatorStaticRef">
+    <dyn:staticRef var="SVarC_PInjPu" staticVar="p"/>
+    <dyn:staticRef var="SVarC_PInjPu" staticVar="q"/>
```
Comment (cossartque) [inline tests/outputs/reference/TestDydDynModel/TestDydDynModel.dyd:40]: Same

File: tests/outputs/reference/TestDydHvdc/TestDydHvdc.dyd [hunk 10]
```diff
@@ -24,6 +27,12 @@
     <dyn:staticRef var="QPu_value" staticVar="q"/>
     <dyn:staticRef var="state_value" staticVar="state"/>
   </dyn:macroStaticReference>
+  <dyn:macroStaticReference id="StaticVarCompensatorStaticRef">
+    <dyn:staticRef var="SVarC_PInjPu" staticVar="p"/>
+    <dyn:staticRef var="SVarC_PInjPu" staticVar="q"/>
```
Comment (cossartque) [inline tests/outputs/reference/TestDydHvdc/TestDydHvdc.dyd:32]: same

File: tests/outputs/reference/TestDydRemote/TestDydRemote.dyd [hunk 11]
```diff
@@ -24,6 +27,12 @@
     <dyn:staticRef var="QPu_value" staticVar="q"/>
     <dyn:staticRef var="state_value" staticVar="state"/>
   </dyn:macroStaticReference>
+  <dyn:macroStaticReference id="StaticVarCompensatorStaticRef">
+    <dyn:staticRef var="SVarC_PInjPu" staticVar="p"/>
+    <dyn:staticRef var="SVarC_PInjPu" staticVar="q"/>
```
Comment (cossartque) [inline tests/outputs/reference/TestDydRemote/TestDydRemote.dyd:32]: same

**Review by cossartque — CHANGES_REQUESTED** [review]

**Review by cossartque — APPROVED** [review]

**Review by marcochiaramello — APPROVED** [review]

**Comment by lecourtoisflo** [comment]
This PR should be merged before https://github.com/dynawo/dynaflow-launcher/pull/197 as it will be easier to solve conflicts between the two

File: sources/Outputs/include/Par.h [hunk 12]
```diff
@@ -84,6 +88,7 @@ class Par {
     const algo::ShuntCounterDefinitions& shuntCounters;                           ///< Shunt counters to use
     const algo::DynamicModelDefinitions& dynamicModelsDefinitions;                ///< list of defined dynamic models
     const algo::LinesByIdDefinitions& linesByIdDefinitions;                       ///< lines by id to use
+    algo::StaticVarCompensatorDefinitions svarcsDefinitions;                      ///< the SVarC definitions to use
```
Comment (gautierbureau) [inline sources/Outputs/include/Par.h:91]: why this one not a const Obejc&  like the others ?

Comment (lecourtoisflo) [reply sources/Outputs/include/Par.h]: Corrected

**Pull request merged**